rmw_zenoh_cpp/src/detail/payload.hpp

```diff
@@ -0,0 +1,88 @@
+// Copyright 2024 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DETAIL__PAYLOAD_HPP_
+#define DETAIL__PAYLOAD_HPP_
+
+#include <zenoh.h>
+
+#include <variant>
+#include <utility>
+#include <vector>
+
+namespace rmw_zenoh_cpp
+{
+///=============================================================================
+class Payload
+{
+public:
+  explicit Payload(const zenoh::Bytes & bytes)
+  {
+    auto slices = bytes.slice_iter();
+    auto slice = slices.next();
+    if (!slice.has_value()) {
+      bytes_ = nullptr;
+    } else {
+      if (!slices.next().has_value()) {
+        bytes_ = Contiguous {slice.value(), bytes.clone()};
```

Comment:
So I'm looking at the definition of `zenoh::Slice` in https://github.com/eclipse-zenoh/zenoh-cpp/blob/e8eca99b4eaff0963e52aefd8405909c1552080d/include/zenoh/api/bytes.hxx#L44-L47 .  It looks to me like this is a shallow copy, and thus we are just copying the `data` pointer and `len` here.  Is this safe to do?  How are we guaranteeing that the data that is pointed to by the `data` pointer still exists later on when we try to access it?

Reply:
That is precisely why we keep a clone of `bytes`, ensuring that the `Bytes` reference count maintained by Zenoh wouldn't go to zero and invalidate the slice.

Reply:
Right, but what's the relationship between the pointer returned by `bytes.clone()`, and `slice.data`?  It looks to me like in the `Contiguous` case, we clone the bytes, but then return the data from the `slice.data` later anyway.

Reply:
These are roughly the definitions on the Rust side:

```rust
pub struct ZBytes(ZBuf);

pub struct ZBuf {
    slices: SingleOrVec<ZSlice>,
}

pub struct ZSlice {
    buf: Arc<dyn ZSliceBuffer>,
    start: usize,
    end: usize,
    #[cfg(feature = "shared-memory")]
    pub kind: ZSliceKind,
}
```

So `Bytes` is roughly a list of reference-counted buffers. We only need to maintain a non-zero the reference count of the `Arc` to ensure that `slice.data` stays valid, as it points directly to the buffer behind the `Arc`.

See:
- https://github.com/eclipse-zenoh/zenoh/blob/a6bc509b5997c922a56253455149858af715edfa/commons/zenoh-buffers/src/zslice.rs#L92C1-L98C2
- https://github.com/eclipse-zenoh/zenoh/blob/a6bc509b5997c922a56253455149858af715edfa/commons/zenoh-buffers/src/zbuf.rs#L32-L34
- https://github.com/eclipse-zenoh/zenoh/blob/a6bc509b5997c922a56253455149858af715edfa/zenoh/src/api/bytes.rs#L93

Reply:
Thanks for the explanation.

From what I can tell of your explanation, then, we need to hold onto the `zenoh::Bytes` object as the reference count, and then that keeps the `zenoh::Slice` buffer around until we drop the reference later.  Do I have that right?

Assuming that is the case, I'm going to suggest adding a comment to that effect, as it is very non-obvious in this code.  It might even be worth renaming the `bytes` member of the structure to `ref` as well, but I'll leave that up to you.

Reply:
Please see if the comment I added in e76073831c078012343e315cf3564be68fa02489 is satisfactory.

Reply:
That comment looks great, though I think it got lost in the recent refactor.